### PR TITLE
feat: Support remote scripts with `uv run`

### PR DESCRIPTION
**Pull request by manzt**
First off, congratulations on the 0.3 release! The PEP 723 standalone scripts support is awesome, and I can already imagine a long tail of little scripts of my own that would benefit from this functionality.

## Background

I really like the Deno CLI's support for running and installing remote scripts. 

```
deno run <url>
```

```
deno install --name foo <url>
```

I can see parallels with `uv run` and `uvx`. After mentioning this on Discord, @zanieb suggested I could take a stab at a PR to implement similar functionality for uv.

## Summary

This PR attempts to add support for executing remote standalone scripts directly with `uv run`. While this is already possible by downloading the script (i.e., via curl/wget) and then using uv run, having direct support would be convenient.

The proposed functionality is:

```sh
uv run <url>
```

Another addition/alternative could be to support running scripts via stdin:

```sh
curl -sL <url> | uv run -
```

But that is not implemented in this PR.

## Test Plan

I noticed that GitHub and `files.pythonhosted.org` URLs are used in some of the tests. I've created a personal [GitHub Gist](https://gist.github.com/manzt/cb24f3066c32983672025b04b9f98d1f) with the example from PEP 723 for now to test this functionality.

~However, I couldn't figure out how to get the `with_snapshot` config filter to filter out the tempfile path, so the test is currently failing. Any assistance with this would be appreciated.~

## Notes

I'm not totally pleased with the implementation of this PR. I think it would be better to handle the case earlier (and probably reuse the cache), and avoid mutation, but since run command requires a local path this was the simplest implementation I could come up with.

I know that performance is paramount with uv so I totally understand if this requires a different approach or something more explicit to avoid "inferring" the path. I'm just taking this as an opportunity to learn a little more Rust and acquaint myself with the code base. cheers!

### CODE REVIEW

**Comment by zanieb** [comment]
Awesome! Thanks for putting this up.

File: crates/uv/tests/run.rs
Comment (manzt) [inline crates/uv/tests/run.rs:1380]: Just copied the example from [PEP 723](https://peps.python.org/pep-0723/), but realizing now the snapshot will likely change over time (with new versions of deps). Probably better to have a more simple test / and simple dependency.

Comment (zanieb) [reply crates/uv/tests/run.rs:1380]: We'll probably create something in the Astral org anyway, thanks!

**Comment by mgaitan** [comment]
Amazing feature. Imagine this with #6283

**Comment by zanieb** [comment]
Sorry for the delay, I've been stretched thin on reviews.

@BurntSushi would you be able to take over review of this?

**Comment by zanieb** [comment]
I generally have a preference for the "We could look for a file path matching the URL, and if it exists, either use the local file or report an error." option rather than prompting. Opt-in might make sense, but via a `--remote` flag rather than a persistent configuration option (maybe that's what you meant though).

**Comment by zanieb** [comment]
See also https://github.com/astral-sh/uv/issues/7396#issuecomment-2351129628

**Comment by BurntSushi** [comment]
If we go the route of prioritizing a local file when given a URL when a file exists, then I think that probably satisfies me. I can't think of a way for that to go wrong from a "accidentally executed a remote script instead of the intended local file" perspective.

There may still be an argument in favor of a `--remote` flag, but more in a vague "otherwise it's too easy to run remote scripts" sense. I don't feel as strongly about that personally.

@manzt What do you think?

**Comment by manzt** [comment]
Thank you for raising this concern. I agree it's a valid issue I hadn't considered.

My preference aligns with @zanieb's suggestion to prioritize checking for a local file path matching the URL first, erroring if a match is found. For context, I checked your example with deno, which always makes the HTTP request, though it operates in a more sandboxed environment than Python.

Regarding a `--remote` flag, while I don't have strong arguments for or against it, I appreciate its explicitness. The common use case I envision is sharing links to Gists or GitHub, where adding `--remote` before execution seems a reasonable safeguard (especially if the readme just has a copy link).

It's worth noting that with #6467, I believe scripts can already be executed via `curl <url> | uv run -`. However, handling this directly in uv does seem really nice UX. 

Apologies for the delayed response!

**Comment by BurntSushi** [comment]
All righty, let's go with checking for whether a local file exists or not.

@charliermarsh Do you have any opinions here? Do you think `uv run https://example.com/foo.py` should "just work," or do you think there should be an extra step needed, e.g., `uv run --remote https://example.com/foo.py`? (Above we discussed mitigating the case where `https://example.com/foo.py` actually refers to a valid file path, in which case, we would just read the local file instead.)

**Comment by BurntSushi** [comment]
Same.

**Comment by charliermarsh** [comment]
Same. Whatever we do, it would be nice to align the behavior of requirements.txt files, which can also be either remote or local.

**Comment by manzt** [comment]
Seems like agreement, I’ll start addressing the prior comments and handle the file behavior!

**Comment by manzt** [comment]
Thanks for your patience. I think I was able to achieve the desired behavior and added another test.

I'm not sure how we should plan on testing the remote behavior (right now still depends on a gist of my own).

File: crates/uv/src/lib.rs
Comment (BurntSushi) [inline crates/uv/src/lib.rs:172]: Why does this start with an underscore? I think you should be able to remove it.

File: crates/uv/src/lib.rs
Comment (BurntSushi) [inline crates/uv/src/lib.rs:58]: I think you should be using [`Path::try_exists`](https://doc.rust-lang.org/std/path/struct.Path.html#method.try_exists) here. If it's result is anything other than `Ok(true)`, then you bail.

The idea here is that I think we want to be pretty conservative: we should only execute a remote script if we can _absolutely confirm_ there is no corresponding local file path. Otherwise, you open yourself up to other vectors where by an end user unintentionally runs a remote script.

File: crates/uv/src/lib.rs
Comment (BurntSushi) [inline crates/uv/src/lib.rs:56]: What happens if there is more than one positional argument? I think this means we ignore it? Should we return `Ok(None)` instead? (I'm actually not quite sure about this one.)

File: crates/uv/src/lib.rs
Comment (BurntSushi) [inline crates/uv/src/lib.rs:63]: I would use an early return here. So if the target _doesn't_ look like a URL, then `return Ok(None);`. This keeps rightward drift in check and makes it easy to see under what conditions the function bails.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
Nice work! This is coming along. :-) I've requested another round of changes and suggested a way forward with respect to testing.

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs:177]: Must Rust knowledge isn't that great, but it seems the `Drop` implementation for `NamedTempFile` unlinks the temporary file. MRE:

```rust
use tempfile::NamedTempFile;

fn main() {
    let path = {
        let tmpfile = NamedTempFile::new().unwrap();
        tmpfile.path().to_str().unwrap().to_string()
    };
    assert!(std::path::Path::new(&path).exists()); // panics!
}
```

The only way I could figure out how to keep it around is to have `resolve_script_target` return the owned temp file and hoist it into a higher scope (then clippy yelled at me for an unused variable, hence the underscore).

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs:47]: Awesome, thanks for taking the time to explain that.

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs]: > I think you should be using `Path::try_exists` here.

Thanks for the pointer!

> If it's result is anything other than `Ok(true)`, then you bail.

Maybe I'm having a hard time thinking this through, but right we "bail" from this function if a file exists. I think we want to bail for anything that is not `Ok(false)`? (I went ahead and made these changes, sorry if I'm missing something).

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs]: Good idea.

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs:57]: > What happens if there is more than one positional argument?

if there is more than one, we don't change them (so they are forwarded down to the python program). E.g., it would be nice to support:

```sh
uv run <url> [python program flags]
```

File: crates/uv/tests/run.rs
Comment (manzt) [inline crates/uv/tests/run.rs:2139]: should we point to main here?

**Comment by manzt** [comment]
hmm 

```sh
cargo run run https://raw.githubusercontent.com/astral-sh/uv/main/scripts/uv-run-remote-script-test.py CI
```

seems to be working on this branch, but the test suite is having trouble resolving the version of `rich`. Maybe I'm missing something in the setup of the `TestContext`.

File: crates/uv/tests/run.rs
Comment (BurntSushi) [reply crates/uv/tests/run.rs:2139]: No I think the commit revision like you have it here is the right way to go. But I would use the full revision hash.

**Comment by BurntSushi** [comment]
> hmm
> 
> ```shell
> cargo run run https://raw.githubusercontent.com/astral-sh/uv/main/scripts/uv-run-remote-script-test.py CI
> ```
> 
> seems to be working on this branch, but the test suite is having trouble resolving the version of `rich`. Maybe I'm missing something in the setup of the `TestContext`.

Ah right my bad! Our tests set `--exclude-newer` to avoid new releases of packages changing resolutions. Really sorry (especially since I'm the one who asked for a version pin!), but this means we'll need another PR to change the script we just merged to use an older version of `rich`. I think `13.7.0` should be old enough since that was published before our `EXCLUDE_NEWER` value:

https://github.com/astral-sh/uv/blob/c8357b7bf23b4ade3b98be74bd7a469c05e6ec02/crates/uv/tests/common/mod.rs#L29-L30

File: crates/uv/src/lib.rs
Comment (BurntSushi) [reply crates/uv/src/lib.rs:57]: Perfect. SGTM.

File: crates/uv/src/lib.rs
Comment (BurntSushi) [inline crates/uv/src/lib.rs:63]: I think you can just do:

```rust
if !matches!(Path::new(target).try_exists(), Ok(false)) {
    return Ok(None);
}
```

There are actually a lot of different ways to write this, but I think I like this one the best. But the `if` with the empty body is a bit odd. :)

**Comment by manzt** [comment]
No worries, I opened https://github.com/astral-sh/uv/pull/7713

**Comment by manzt** [comment]
Test is working now, pointing to #7713. Will switch to hash on main once merged.

File: crates/uv/src/lib.rs
Comment (BurntSushi) [inline crates/uv/src/lib.rs:80]: Will this result in `script.py.py` if the `script.py` fallback is used above?

File: crates/uv/src/lib.rs
Comment (BurntSushi) [reply crates/uv/src/lib.rs:177]: Ahhhh yes okay, I didn't realize it wasn't actually being used below. This is kind of interesting because I'm not sure there is actually a guarantee that `_maybe_tempfile` won't just be dropped right away. The way to fix that for sure would be this patch:

```rust
diff --git a/crates/uv/src/lib.rs b/crates/uv/src/lib.rs
index 6c21f243..c6096124 100644
--- a/crates/uv/src/lib.rs
+++ b/crates/uv/src/lib.rs
@@ -174,10 +174,10 @@ async fn run(mut cli: Cli) -> Result<ExitStatus> {
     };
 
     // Parse the external command, if necessary.
-    let mut _maybe_tempfile: Option<tempfile::NamedTempFile> = None;
+    let mut maybe_tempfile: Option<tempfile::NamedTempFile> = None;
     let run_command = if let Commands::Project(command) = &mut *cli.command {
         if let ProjectCommand::Run(uv_cli::RunArgs { command, .. }) = &mut **command {
-            _maybe_tempfile = resolve_script_target(command).await?;
+            maybe_tempfile = resolve_script_target(command).await?;
             Some(RunCommand::try_from(&*command)?)
         } else {
             None
@@ -298,7 +298,7 @@ async fn run(mut cli: Cli) -> Result<ExitStatus> {
     // Configure the cache.
     let cache = Cache::from_settings(cache_settings.no_cache, cache_settings.cache_dir)?;
 
-    match *cli.command {
+    let result = match *cli.command {
         Commands::Help(args) => commands::help(
             args.command.unwrap_or_default().as_slice(),
             printer,
@@ -1185,7 +1185,9 @@ async fn run(mut cli: Cli) -> Result<ExitStatus> {
                 commands::build_backend::prepare_metadata_for_build_editable(&wheel_directory)
             }
         },
-    }
+    };
+    drop(maybe_tempfile);
+    result
 }
 
 /// Run a [`ProjectCommand`].
```

But this brings up an interesting point: what happens if executing the downloaded Python script fails? I guess it would show the temporary file path and not the URL, which is potentially confusing.

File: crates/uv/src/lib.rs
Comment (BurntSushi) [reply crates/uv/src/lib.rs]: Yes, you're 100% right. I got the logic flipped around! Whoops.

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs:63]: ooo, I like this.

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs]: I applied the patch here. 

> But this brings up an interesting point: what happens if executing the downloaded Python script fails? I guess it would show the temporary file path and not the URL, which is potentially confusing.

There is already some indication of the tempfile location with the summary of parsing script 723 metadata:

```sh
    Reading inline script metadata from: [TEMP_PATH].py
```

Perhaps we could add some context/logging that says "Creating a tempfile for [URL] in [TEMP_PATH]?

thus tracebacks from a failed Python script would have more context?

File: crates/uv/tests/run.rs
Comment (manzt) [reply crates/uv/tests/run.rs]: Done.

File: crates/uv/src/lib.rs
Comment (BurntSushi) [reply crates/uv/src/lib.rs]: > Perhaps we could add some context/logging that says "Creating a tempfile for [URL] in [TEMP_PATH]?

I like this!

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs:79]: Ah yeah, you are right. tempfile adds a hash and a suffix to the prefix.

```sh
cargo run run https://raw.githubusercontent.com/astral-sh/uv/main/scripts/uv-run-remote-script-test.py CI
# Reading inline script metadata from: /var/folders/gs/j5y7yv816z71dzy18r4qmy3h0000gq/T/uv-run-remote-script-test.pyznxLtC.py
# Installed 4 packages in 21ms
# Hello CI, from uv!
```

It might be easiest to:

```rust
    let temp_file = tempfile::Builder::new()
        .suffix(".py")
        .tempfile?;
```

Depends on how useful it is to try for the file name. It's only used to report to the end user and would show up in traceback messages from Python errors.

Comment (BurntSushi) [reply crates/uv/src/lib.rs:79]: I think including the file name in there as a prefix is fine.  Especially if we add a log message saying that the remote script was written to the temp file.

I think my issue here is the case where the `script.py` _fallback name_ is used (from above). Does that result in a double `.py` suffix?

Comment (manzt) [reply crates/uv/src/lib.rs:79]: I think the issue also happens if the file has as `.py`. For example:

https://raw.githubusercontent.com/astral-sh/uv/main/scripts/uv-run-remote-script-test.py

results in:

uv-run-remote-script-test.pyznxLtC.py

So `script.py` would similarly result in `script.py[hash].py`.

Comment (BurntSushi) [reply crates/uv/src/lib.rs:79]: Yeah, so I think you want to strip the `.py` suffix from the file name, but keep the `.py` suffix in the temp file builder.

Comment (manzt) [reply crates/uv/src/lib.rs:80]: Ah, much more simple. Done!

**Comment by BurntSushi** [comment]
The CI failure looks legitimate. My guess is that the file path is just totally invalid on Windows, which in turn causes `Path::new(...).try_exists()` to fail and thus bail out of the remote logic. So I think I would change the `if` to this:

```rust
    // Only continue if we are absolutely certain no local file exists.
    //
    // We don't do this check on Windows since the file path would
    // be invalid anyway, and thus couldn't refer to a local file.
    if cfg!(unix) && !matches!(Path::new(target).try_exists(), Ok(false)) {
        return Ok(None);
    }
```

Also, I think we need to add docs to `uv run` mentioning this feature. :)

**Comment by manzt** [comment]
Thanks for the tips!

> Also, I think we need to add docs to uv run mentioning this feature. :)

I took a stab at updating the docs.

**Comment by BurntSushi** [comment]
I think you need to run `cargo dev generate-cli-reference`. Basically, changes to the docs need to be propagated out to other places.

**Comment by manzt** [comment]
Ah, sorry... wasn't sure what the source of truth was. Thanks!

**Comment by manzt** [comment]
Any thoughts on how to update the test checking for the local file? Is there a way to mark that a test _should_ fail on windows for example?

File: crates/uv/src/lib.rs
Comment (BurntSushi) [inline crates/uv/src/lib.rs:89]: Hmmmm. This is an issue I think. We should be using a `Printer` here derived from config settings. Otherwise this message won't be squashed when, e.g., `--quiet` is passed. I'm not quite sure how to solve that. Another approach would be to emit this message after a `Printer` is constructed below. See:

```
    // Configure the `Printer`, which controls user-facing output in the CLI.
    let printer = if globals.quiet {
        Printer::Quiet
    } else if globals.verbose > 0 {
        Printer::Verbose
    } else if globals.no_progress {
        Printer::NoProgress
    } else {
        Printer::Default
    };
```

File: docs/reference/cli.md
Comment (BurntSushi) [inline docs/reference/cli.md:57]:
```suggestion
When used with a file ending in `.py` or an HTTP(S) URL, the file will be treated as a script and run with a Python interpreter, i.e., `uv run file.py` is equivalent to `uv run python file.py`. For URLs, the script is downloaded to a temporary file before execution. If the script contains inline dependency metadata, it will be installed into an isolated, ephemeral environment. When used with `-`, the input will be read from stdin, and treated as a Python script.
```

**Review by BurntSushi — COMMENTED** [review]
> Any thoughts on how to update the test checking for the local file? Is there a way to mark that a test should fail on windows for example?

I don't think it should fail though? It looks like it's failing because it can't write a temp file? I'm not sure what the issue is there.

File: crates/uv/src/lib.rs
Comment (manzt) [reply crates/uv/src/lib.rs]: I updated the script to check whether the tempfile exists and report that a file was downloaded (reusing the `Printer`).

I think this is the most minimal way to respect `--quiet`. From an end user perspective, it would probably be nicer to report when we are actually starting to download the script, but I'm not quite should how to solve that without a bigger refactor.

One idea would be to create the temp file immediately and kick off the download, but await a future for the download/writing after writing "Downloading remote script...". Maybe too complicated.

**Comment by manzt** [comment]
> I don't think it should fail though? It looks like it's failing because it can't write a temp file? I'm not sure what the issue is there.

I think on Windows we skip checking for a local file, which would result in requesting`https://example.com/path/to/main.py`

> It looks like it's failing because it can't write a temp file?

Right, I'm also not sure of this error... but if we were to write the tempfile I think it would still fail. Maybe I'm missing something.

**Comment by BurntSushi** [comment]
> I think on Windows we skip checking for a local file, which would result in requestinghttps://example.com/path/to/main.py

OK, so you're referring to the `run_url_like_with_local_file_priority` test here. I was looking at `run_remote_pep723_script`.

For `run_remote_pep723_script`, I think the test should pass.

For `run_url_like_with_local_file_priority`, right, the thing being tested doesn't make sense on Windows. So I would probably just add a `#[cfg(unix)]` on that test with a comment explaining why that's there.

File: crates/uv/src/lib.rs
Comment (BurntSushi) [reply crates/uv/src/lib.rs]: I think what you have is fine for now. Thanks!

**Comment by manzt** [comment]
> OK, so you're referring to the `run_url_like_with_local_file_priority` test here. I was looking at `run_remote_pep723_script`.

Yes, apologies for the confusion. I added the `#[cfg(unix)]` with a comment.

**Comment by manzt** [comment]
> For run_remote_pep723_script, I think the test should pass.

Agreed.

**Comment by manzt** [comment]
I wonder if the issue on Windows is related to some behavior behind how [`NamedTempFile`](https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#windows) works on Windows.

**Comment by BurntSushi** [comment]
> I wonder if the issue on Windows is related to some behavior behind how [`NamedTempFile`](https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#windows) works on Windows. Are we "Opening a named temporary file in a world-writable directory."?

I'm not sure. It seems like it's [`NamedTempFile::persist`](https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#method.persist) that's failing.

Oh, I think I see the problem. You're using `write_atomic` and passing the path of the temporary file you just created to it. And `write_atomic` tries to create another temporary file at that same path and persists it. Specifically, I think this is where the error is coming from:

https://github.com/astral-sh/uv/blob/c07000718aa8eb3321f5d46bd454ae863ffc1963/crates/uv-fs/src/lib.rs#L147-L156

I don't think you need `write_atomic` here. Can you say more about why you used it? I think just a [`write_all`](https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all) on the `NamedTempFile` itself is sufficient.

**Comment by manzt** [comment]
> Oh, I think I see the problem.

Thanks for explanation. I've learned a lot from this PR.

> I don't think you need write_atomic here. Can you say more about why you used it? I think just a write_all on the NamedTempFile itself is sufficient.

It was naive of me. I was looking for how files were written (async) in other parts of the project, but for a tempfile I realize we really don't need. Replaced with `write_all`!

**Comment by charliermarsh** [comment]
Do we need a file for this? Could we just read into memory and pass the contents over stdin?

**Comment by charliermarsh** [comment]
(I could also look into that separately, but I'm wondering if it would remove a lot of the complexity around tempfiles, drop, etc.)

**Comment by manzt** [comment]
> Do we need a file for this? Could we just read into memory and pass the contents over stdin

Definitely! However, I went this direction because writing to a tempfile was the easiest way to support PEP 723 metadata and passing down arguments to the Python script. IIRC PEP 723 is only parsed when cli is given a file path and stdin resolves to `python -c <contents>`, which I wasn't sure how to deal with additional flags.

**Comment by charliermarsh** [comment]
Okay, makes sense. I think we should follow-up by changing that -- it seems like an oversight that we don't detect scripts there!

**Comment by charliermarsh** [comment]
Funnily enough: https://github.com/astral-sh/uv/issues/8097

**Comment by charliermarsh** [comment]
I defer to @BurntSushi to merge.

**Review by BurntSushi — APPROVED** [review]
I think this LGTM now! Amazing work @manzt. Thank you for sticking with this. I know there was a lot of back-and-forth to get this merged, but I think it's in a good state now!

**Comment by BurntSushi** [comment]
I think you just need to rebase this on master.

**Comment by BurntSushi** [comment]
> I think you just need to rebase this on `main`.

Nope that was wrong! I usually do rebase & merge for my PRs, so that was selected (even though I was always planning to squash this PR) and _that_ way greyed out. But squash & merge was fine. Interesting.

**Comment by manzt** [comment]
Ok good! I tried rebasing and it was scary :)

**Comment by hauntsaninja** [comment]
Downloading has some semantic advantages, `__file__` is meaningful, tracebacks are better, etc

**Comment by charliermarsh** [comment]
Yeah I looked into it and agree that downloading is actually better.

**Comment by afeblot** [comment]
Was coming here to suggest/ask for this feature, and found it just released already. I love you :-)

**Pull request merged**
